xarray/backends/common.py

```diff
@@ -373,6 +376,14 @@ class BackendEntrypoint:
       ``filename_or_obj``, ``False`` otherwise. The implementation of this
       method is not mandatory.
 
+    Optionally, it shall implement:
+
+    - ``open_datatree`` method: it shall implement reading from file, variables
+      decoding and it returns an instance of :py:class:`~datatree.DataTree`.
+      It shall take in input at least ``filename_or_obj`` argument and
+      ``drop_variables`` keyword argument.
+      For more details see TODO.
```

Comment:
I wonder if, instead of making `open_datatree` optional, it would be possible to expose a separate entrypoint?

My motivation is a composite backend where a `Dataset` entrypoint would not make sense (I'm using `open_dataset` to open several netcdf files and arrange them in a hierarchy encoded in the file names).

Reply:
@keewis My instinct is that a separate entrypoint would give a messier API (would we need three entrypoints for each combination of datatree and/or dataset implementation, each subclassed from a private base class?). Perhaps there would be a clever way to just have at least one of `open_datatree` or `open_dataset` required, so a single entrypoint (i.e., the existing `BackendEntrypoint`) could handle every combination?

Reply:
> Perhaps there would be a clever way to just have at least one of open_datatree or open_dataset required, so a single entrypoint (i.e., the existing BackendEntrypoint) could handle every combination?

Seems possible, see https://github.com/pydata/xarray/pull/7460#discussion_r1278000861

Reply:
the idea I had was that you'd have two entirely separate entrypoints (as in, package metadata entrypoints that can be registered against), one for `Dataset` / `DataArray` and one for `DataTree`. The advantage would be to keep both entirely separate, so we wouldn't have a need for `open_dataset_parameters` and `open_datatree_parameters`.

However, that might make it a bit confusing since it would allow two different packages to register entrypoints under the same name, so I won't insist on it (which may or may not be intentional). Plus, keeping the functions in a single `BackendEntrypoint` makes raising helpful error messages a bit easier.

(And, one might argue that given any of the functions it is possible to implement the others, although it might not be the most efficient way to implement them)